fdbserver/ApplyMetadataMutation.cpp

```diff
@@ -1257,70 +1191,318 @@ class ApplyMetadataMutationsImpl {
+		size_t size() const { return this->_size; }
+	};
+	Array data;
+
+public:
+	FunctionMap(Array const& data) : data(data) {}
+	struct Builder {
+		Arena& arena;
+		std::vector<std::pair<KeyRangeRef, ApplyFunctions>> functions;
+
+		Builder(Arena& arena) : arena(arena) {}
+
+		void insert(KeyRangeRef range, ApplyFunctions&& fun) { functions.emplace_back(range, std::move(fun)); }
+
+		// building this function map is expensive and probably could be optimized better. However, this code is
+		// executed only once in the lifetime of a proxy and it works on a small set of values. So we trade performance
+		// to build this map for lookup performance (the result will be a read-only map)
+		std::unique_ptr<FunctionMap> done() {
+			std::sort(functions.begin(), functions.end(), [](auto const& lhs, auto const& rhs) {
+				return lhs.first.begin < rhs.first.begin;
+			});
+			auto maxIter = std::max_element(functions.begin(), functions.end(), [](auto const& lhs, auto const& rhs) {
+				return lhs.first.begin.size() < rhs.first.begin.size();
+			});
+			uint32_t entrySize = maxIter->first.begin.size() + sizeof(ArrayEntry);
+			// now we can build an array of functions where the strings are all inlined. This will improve the speed of
+			// the binary search when looking up elements
+			// we need to count the number of unique prefixes -- while std::set is not the fastest way of doing this it
+			// is convenient
+			std::set<KeyRef> prefixes;
+			for (auto const& p : functions) {
+				prefixes.insert(p.first.begin);
+			}
+			unsigned elementCount = prefixes.size();
+			Array res(arena, elementCount, entrySize);
+			auto vecIter = functions.begin();
+			for (auto iter = res.begin(); iter != res.end(); ++iter) {
+				// each element starts with an array entry and then has enough memory so we can copy the begin string
+				// into it. While this is not strictly necessary (everything would work without doing this), it will
+				// improve memory locality and therefore speed up binary search
+				uint8_t* strCopy = iter.ptr + sizeof(ArrayEntry);
+				memcpy(strCopy, vecIter->first.begin.begin(), vecIter->first.begin.size());
+				iter->begin = KeyRef(strCopy, vecIter->first.begin.size());
+				iter->value.end = vecIter->first.end;
+				iter->value.fun = vecIter->second;
+				iter->value.next = nullptr;
+				++vecIter;
+				// add functions with the same key prefix
+				for (; vecIter != functions.end() && vecIter->first.begin == iter->begin; ++vecIter) {
+					iter->value.next = new (arena) ValueType{ .end = vecIter->first.end, .fun = vecIter->second };
+				}
+			}
+			ASSERT(vecIter == functions.end());
+			return std::make_unique<FunctionMap>(res);
 		}
+	};
 
-		for (KeyRangeRef& range : tssServerListToRemove) {
-			txnStateStore->clear(range);
+	void apply(ApplyMetadataMutationsImpl* self, MutationRef m) {
+		if (!isMetadataMutation(m)) {
+			return;
+		}
+		auto iter = std::lower_bound(data.begin(), data.end(), m.param1);
```

Comment:
If iter points to the first element >= m.param1, does that mean we miss actions that we should take for a keyrange kr containing m.param1 where kr.begin < m.param1?